dispatch/group_test.go

```diff
@@ -0,0 +1,351 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dispatch
+
+import (
+	"context"
+	"reflect"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/promslog"
+)
+
+func TestAggrGroup(t *testing.T) {
+	lset := model.LabelSet{
+		"a": "v1",
+		"b": "v2",
+	}
+	opts := &RouteOpts{
+		Receiver: "n1",
+		GroupBy: map[model.LabelName]struct{}{
+			"a": {},
+			"b": {},
+		},
+		GroupWait:      1 * time.Second,
+		GroupInterval:  300 * time.Millisecond,
+		RepeatInterval: 1 * time.Hour,
+	}
+	route := &Route{
+		RouteOpts: *opts,
+	}
+
+	var (
+		a1 = &types.Alert{
+			Alert: model.Alert{
+				Labels: model.LabelSet{
+					"a": "v1",
+					"b": "v2",
+					"c": "v3",
+				},
+				StartsAt: time.Now().Add(time.Minute),
+				EndsAt:   time.Now().Add(time.Hour),
+			},
+			UpdatedAt: time.Now(),
+		}
+		a2 = &types.Alert{
+			Alert: model.Alert{
+				Labels: model.LabelSet{
+					"a": "v1",
+					"b": "v2",
+					"c": "v4",
+				},
+				StartsAt: time.Now().Add(-time.Hour),
+				EndsAt:   time.Now().Add(2 * time.Hour),
+			},
+			UpdatedAt: time.Now(),
+		}
+		a3 = &types.Alert{
+			Alert: model.Alert{
+				Labels: model.LabelSet{
+					"a": "v1",
+					"b": "v2",
+					"c": "v5",
+				},
+				StartsAt: time.Now().Add(time.Minute),
+				EndsAt:   time.Now().Add(5 * time.Minute),
+			},
+			UpdatedAt: time.Now(),
+		}
+	)
+
+	var (
+		last       = time.Now()
+		current    = time.Now()
+		lastCurMtx = &sync.Mutex{}
+		alertsCh   = make(chan types.AlertSlice)
+	)
+
+	ntfy := func(ctx context.Context, alerts ...*types.Alert) bool {
+		// Validate that the context is properly populated.
+		if _, ok := notify.Now(ctx); !ok {
+			t.Errorf("now missing")
+		}
+		if _, ok := notify.GroupKey(ctx); !ok {
+			t.Errorf("group key missing")
+		}
+		if lbls, ok := notify.GroupLabels(ctx); !ok || !reflect.DeepEqual(lbls, lset) {
+			t.Errorf("wrong group labels: %q", lbls)
+		}
+		if rcv, ok := notify.ReceiverName(ctx); !ok || rcv != opts.Receiver {
+			t.Errorf("wrong receiver: %q", rcv)
+		}
+		if ri, ok := notify.RepeatInterval(ctx); !ok || ri != opts.RepeatInterval {
+			t.Errorf("wrong repeat interval: %q", ri)
+		}
+
+		lastCurMtx.Lock()
+		last = current
+		// Subtract a millisecond to allow for races.
+		current = time.Now().Add(-time.Millisecond)
+		lastCurMtx.Unlock()
+
+		alertsCh <- types.AlertSlice(alerts)
+
+		return true
+	}
+
+	removeEndsAt := func(as types.AlertSlice) types.AlertSlice {
+		for i, a := range as {
+			ac := *a
+			ac.EndsAt = time.Time{}
+			as[i] = &ac
+		}
+		return as
+	}
+
+	// Test regular situation where we wait for group_wait to send out alerts.
```

Comment:
This test function covers several scenarios. Would it be better to wrap each scenario in a t.Run() block? That would allow a single failing test to be run repeatedly for easier debugging.